designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,372 @@
+      2. [Fully Blocking Drift & Expiration while Allowing Consolidation](#fully-blocking-drift--expiration-while-allowing-consolidation)
+      3. [Fully Blocking Drift and Expiration while Allowing Emptiness](#fully-blocking-drift-and-expiration-while-allowing-emptiness)
+      4. [Limiting Drift & Expiration while Allowing Aggressive but Not Unbounded Consolidation](#limiting-drift--expiration-while-allowing-aggressive-but-not-unbounded-consolidation)
+      5. [Limiting Cost Saving Disruption to Occur During Weekends](#limiting-cost-saving-disruption-to-occur-during-weekends)
+      6. [Which Equation Best Covers All the Scenarios?](#which-equation-best-covers-all-the-scenarios)
+      7. [How Should We Handle an Unspecified Reason When Others are Specified?](#how-should-we-handle-an-unspecified-reason-when-others-are-specified)
+
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. Users want to minimize workload disruptions during business hours but still want to be able to delete empty nodes throughout the day.  That is, empty can run all day, while limiting cost savings and upgrades due to drift to non-business hours only.
+
+See Less Made Up Scenarios here: 
+- https://github.com/kubernetes-sigs/karpenter/issues/924 
+- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
+- https://github.com/kubernetes-sigs/karpenter/issues/672
+
+## Clarifying the requirements and behavior 
+**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
+**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
+**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 
+
+
+# API Design
+## Approach A: Extending the Budget API to specify a reason 
+This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).
+
+### List Approach: Multiple Reasons per Budget - Recommended
+This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
+#### Proposed Spec
+Add a simple field "reasons" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+      // Reasons is a list of disruption reasons. If Reasons is not set, this budget applies to all methods.
```

Comment:
> If Reasons is not set, this budget applies to all methods.

If equation 2 is chosen this should probably be `If Reasons is not set, this budget applies to all methods, which are not overridden.`?